src/awkward/_nplikes/array_module.py

```diff
@@ -172,6 +174,11 @@ def div_shape_item(self, x1: ShapeItem, x2: ShapeItem) -> ShapeItem:
         assert result * x2 == x1
         return result
 
+    def shape_equals(
+        self, shape1: tuple[ShapeItem, ...], shape2: tuple[ShapeItem, ...]
+    ) -> bool:
+        return shape1 == shape2
+
```

Comment:
New shape comparison function. We use `None` to indicate "unknown". We already have 0-D `TypeTracerArray`s for unknown scalars, but I think this is distinct from arrays as it is a metadata operation. As such, we repurpose `None` (it now means no-length OR unknown-shape/length equality)